zerver/lib/api.py

```diff
@@ -0,0 +1,389 @@
+import inspect
+from dataclasses import dataclass
+from enum import Enum, auto
+from functools import wraps
+from typing import Callable, Generic, List, Mapping, Optional, Sequence, Set, Type, TypeVar, Union
+
+from django.http import HttpRequest
+from django.utils.translation import gettext as _
+from pydantic import Json, TypeAdapter, ValidationError, ValidationInfo
+from typing_extensions import (
+    Annotated,
+    Concatenate,
+    ParamSpec,
+    get_args,
+    get_origin,
+    get_type_hints,
+)
+
+from zerver.lib.exceptions import JsonableError
+from zerver.lib.request import RequestNotes, RequestVariableMissingError, arguments_map
+from zerver.lib.response import MutableJsonResponse
+
+T = TypeVar("T")
+ParamT = ParamSpec("ParamT")
+ReturnT = TypeVar("ReturnT")
+
+
+class DocumentationStatus(Enum):
+    DOCUMENTED = auto()
+    INTENTIONALLY_UNDOCUMENTED = auto()
+    DOCUMENTATION_PENDING = auto()
+
+
+@dataclass(frozen=True)
+class ApiParamConfig:
+    whence: Optional[str] = None
+    path_only: bool = False
+    argument_type_is_body: bool = False
+    documentation_status: DocumentationStatus = DocumentationStatus.DOCUMENTED
+    # This can be used to override Pydantic's configuration to use strict mode.
+    # For endpoint parameters wrapped in pydantic.Json, we default it to True,
+    # otherwise it is default to None to use Pydantic's default.
```

Comment:
This comment might benefit from a few words about what effect it has, or if that's too complex, a link to the documentation, just to make it easy to follow.

(Probably ultimately we'll want to document all the flags here concisely and clearly, since it's a bit of the API)